openedx/course-discovery
# perf: fix N+1 and some duplicate queries issues across CatalogSerializer

### [PROD-3826](https://2u-internal.atlassian.net/browse/PROD-3826)

### Description
- Fix N+1 issues in Catalog serializer
- Use courseurlslug to filter active url slug to avoid altering prefetch results
- Use SlugRelatedFieldwithReadSerializer for product source to avoid duplicate source calls in to_representation
- Add caching in  active_url_slug so that multiple calls to property in same call does not lead to repeated DB calls
- Change course run status handling to get statuses in Python Code instead of ORM to avoid disrupting Prefetch query results
- Fix the tests that have their query count decreased

### Testing
- Checkout master
- Go to admin http://localhost:18381/admin/catalogs/catalog/ and add 1-2 catalogs (In one query, use ""{
"content_type":"course", \n
"partner":"edx",\n
}"" and in another catalog query, use *). 
- Go to http://localhost:18381/api/v1/catalogs/2/courses/ or http://localhost:18381/api/v1/catalogs/1/courses/ (1, 2 are PK of catalog) and note the query count.
- Checkout this branch and hit the urls again. Note the new query count. Hit the url again and there should be small deduction in query count (due to caching). 

### Screenshots

#### Master
<img width="1680" alt="image" src="https://github.com/openedx/course-discovery/assets/40599381/e264b8de-e833-4505-9aac-c6cc0f391273">

#### N+1 fixups
<img width="1680" alt="image" src="https://github.com/openedx/course-discovery/assets/40599381/326dfc58-7e07-40e7-89cc-78f5257ff295">
